collector/src/test/java/com/google/cloud/trace/zipkin/StackdriverTraceClassRule.java

```diff
@@ -0,0 +1,149 @@
+package com.google.cloud.trace.zipkin;
+
+import com.google.common.collect.Sets;
+import com.google.devtools.cloudtrace.v1.PatchTracesRequest;
+import com.google.devtools.cloudtrace.v1.Trace;
+import com.google.devtools.cloudtrace.v1.TraceServiceGrpc;
+import com.google.devtools.cloudtrace.v1.TraceSpan;
+import com.google.protobuf.Empty;
+import io.grpc.Server;
+import io.grpc.netty.NettyServerBuilder;
+import io.grpc.stub.StreamObserver;
+import org.junit.rules.MethodRule;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.FrameworkMethod;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.SocketUtils;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+
+import static java.util.Collections.unmodifiableSet;
+
+/**
+ * Starts up a local Stackdriver Trace server, listening for GRPC requests on {@link #grpcURI}.
+ *
+ * Works with both {@link org.junit.ClassRule} and {@link org.junit.Rule}
+ */
+public class StackdriverTraceClassRule extends TraceServiceGrpc.TraceServiceImplBase implements TestRule, MethodRule
+{
+  private static final Logger LOG = LoggerFactory.getLogger(StackdriverTraceClassRule.class);
+
+  private final int port;
+  private final Server server;
+
+  private final Set<String> traceIds = Sets.newConcurrentHashSet();
+  private final Set<Long> spanIds = Sets.newConcurrentHashSet();
+  private CountDownLatch spanCountdown;
+
+  public StackdriverTraceClassRule()
+  {
+    this.port = SocketUtils.findAvailableTcpPort();
+    this.server = NettyServerBuilder.forPort(port).addService(this).build();
+  }
+
+  @Override
+  public Statement apply(final Statement base, Description description)
+  {
+    return apply(base, null, null);
+  }
+
+  @Override
+  public Statement apply(final Statement base, FrameworkMethod method, Object target)
+  {
+    return new Statement()
+    {
+      @Override
+      public void evaluate() throws Throwable
+      {
+        if (isRunning())
```

Comment:
unless I'm missing something... I think you should trust junit. Let it manage the scope.. it shouldn't attempt to start it per test unless it is configured as a regular rule (as opposed to a class rule). This should make the code less complex too